🔄 Refactor OpenAI function calling and add extension method

lemillermicrosoft opened the pull request:
This commit refactors the OpenAI function calling in the Example58_OpenAIFunctionCalling.cs file by introducing a new method CompleteChatWithFunctionsManuallyAsync. It also adds a new extension method GenerateMessageWithFunctionsAsync in the
ChatCompletionExtensions.cs file to generate chat messages using Functions. The changes improve code organization and provide a more convenient way to generate chat messages with Functions.

### Motivation and Context

<!-- Thank you for your contribution to the semantic-kernel repo!
Please help reviewers and future users, providing the following information:
  1. Why is this change required?
  2. What problem does it solve?
  3. What scenario does it contribute to?
  4. If it fixes an open issue, please link to the issue here.
-->

### Description

<!-- Describe your changes, the overall approach, the underlying design.
     These notes will help understanding how your code works. Thanks! -->

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

lemillermicrosoft committed:
🔄 Refactor OpenAI function calling and add extension method

This commit refactors the OpenAI function calling in the
Example58_OpenAIFunctionCalling.cs file by introducing a new method
CompleteChatWithFunctionsManuallyAsync. It also adds a new extension
method GenerateMessageWithFunctionsAsync in the
ChatCompletionExtensions.cs file to generate chat messages using
Functions. The changes improve code organization and provide a more
convenient way to generate chat messages with Functions.

lemillermicrosoft commented on dotnet/src/Connectors/Connectors.AI.OpenAI/ChatCompletionExtensions.cs:
```diff
@@ -0,0 +1,60 @@
+        this IChatCompletion chatCompletion,
+        ChatHistory chat,
+        IKernel kernel,
+        ChatRequestSettings? requestSettings = null,
```
Could make this OpenAIChatRequestSettings instead and add optional parameters like `AutoInvokeFunctions`

SergeyMenshykh commented on dotnet/src/Connectors/Connectors.AI.OpenAI/ChatCompletionExtensions.cs:
```diff
@@ -0,0 +1,60 @@
+    /// <returns>A task representing the generated chat message in string format.</returns>
+    public static async Task<string> GenerateMessageWithFunctionsAsync(
+        this IChatCompletion chatCompletion,
+        ChatHistory chat,
```
It seems the `chat` parameter is not used in the method.

gitri-ms replied:
It's used in line 39:
```
var chatResults = await chatCompletion.GetChatCompletionsAsync(chat, chatRequestSettings, cancellationToken).ConfigureAwait(false);
```

SergeyMenshykh replied:
Opps, I did not notice :)

gitri-ms commented on dotnet/src/Connectors/Connectors.AI.OpenAI/ChatCompletionExtensions.cs:
```diff
@@ -0,0 +1,60 @@
+    {
+        IReadOnlySkillCollection skillCollection = kernel.Skills;
+        OpenAIChatRequestSettings chatRequestSettings = requestSettings as OpenAIChatRequestSettings ?? new();
+        chatRequestSettings.Functions = skillCollection.GetFunctionsView().ToOpenAIFunctions();
```
We would need to make it super clear to the caller that any functions defined in the request settings will be overwritten here.

gitri-ms commented on dotnet/src/Connectors/Connectors.AI.OpenAI/ChatCompletionExtensions.cs:
```diff
@@ -0,0 +1,60 @@
+    /// <param name="cancellationToken">The asynchronous cancellation token (optional).</param>
+    /// <remarks>This extension does not support multiple prompt results (only the first will be returned).</remarks>
+    /// <returns>A task representing the generated chat message in string format.</returns>
+    public static async Task<string> GenerateMessageWithFunctionsAsync(
```
I'm really not a fan of this approach, because I think it creates confusion about what this extension returns. If a function is not invoked, it will return a message. But if a function is invoked, the result might not be a message per se -- it could be a serialized JSON object or some other result that doesn't necessarily readily fit into a chat flow. The caller needs to be able to differentiate between those different outputs.

lemillermicrosoft replied:
Yeah I understand the issue now. This needs to add a chat message as role 'function' to the history and generate a second openai response.

lemillermicrosoft committed:
🔄 Improve function handling and update ChatCompletionExtensions

Refactor the `CompleteChatWithFunctionsAsync` method to return a
`ChatHistory` object instead of a string. Update the method to handle
function responses more efficiently and add new messages to the chat
history. Also, fix the namespace import for `Microsoft.SemanticKernel`.

Update ChatCompletionExtensions to return a list of messages instead of
just the first message content. Add a new AuthorRole for function call
responses in chat completions.

lemillermicrosoft committed:
WIP 🤖 Update OpenAI and chat functionality integration

- Update OpenAI chat model ID in Example59_OpenAIFunctionCalling.cs
- Add ChatRole.Function validation in ClientBase.cs
- Modify ChatCompletionExtensions.cs to handle function calls
- Introduce ChatFunctionMessage in ChatHistory.cs
- Allow nullable content in ChatMessageBase.cs
